File: Android/app/src/main/java/com/auralis/crisisconnect/nearby/NearbySpakePairing.kt
package com.auralis.crisisconnect.nearby

import com.auralis.crisisconnect.security.Crypto
import com.auralis.crisisconnect.security.Spake2P256
import java.io.ByteArrayOutputStream
import java.math.BigInteger
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.util.UUID

/**
 * Harvest-proof offline "add by phone number" via SPAKE2 (RFC 9382), replacing the old passive
 * number-beacon. The phone number is used only as the low-entropy SPAKE2 password, so it is NEVER
 * put on the air: an active attacker gets exactly one guess per live handshake and learns nothing
 * on a wrong guess, so recovering a number would need ~1e9 in-person handshakes with the victim.
 *
 * Flow (initiator = the searcher who has the target's number; responder = the discoverable device,
 * who uses its OWN number). Both derive the same w only if they mean the same number:
 *   1. initiator → responder : pA
 *   2. responder → initiator : pB, cB          (responder's key-confirmation MAC)
 *   3. initiator → responder : cA, enc(identityA)   (verifies cB first, then reveals identity)
 *   4. responder → initiator : status, enc(identityB)  (verifies cA first)
 * On success both hold the SPAKE2 key Ke; identities (uid + messaging public key + name) are
 * exchanged AES-256-GCM-encrypted under a key derived from Ke, so eavesdroppers on the plain BLE
 * link learn nothing and the identity is bound to the number-authenticated session (→ TOFU pin).
 *
 * This crypto/session core is transport-agnostic and unit-tested; the GATT plumbing and the
 * user-consent gate live in the service layer. SPAKE2 itself is verified against the RFC 9382
 * test vector (see Spake2P256Test).
 */
object NearbySpakePairing {
    private const val VERSION = 1
    private const val DOMAIN = "crisisconnect:nearby:spake2:v1"

    // GATT service for the SPAKE2 pairing (distinct from the mesh 0xCCA1 and the retired
    // number-beacon 0xCCA2). The connector WRITEs a message to REQUEST and READs the reply from
    // RESPONSE; the 4-message flow reuses these two characteristics across two write/read rounds.
    val SERVICE_UUID: UUID = UUID.fromString("0000cca6-0000-1000-8000-00805f9b34fb")
    val REQUEST_UUID: UUID = UUID.fromString("0000cca7-0000-1000-8000-00805f9b34fb")
    val RESPONSE_UUID: UUID = UUID.fromString("0000cca8-0000-1000-8000-00805f9b34fb")

    // Fixed SPAKE2 identities (RFC 9382 §3.3 idA/idB). Security comes from w (the number); the peer
    // identities are exchanged inside the authenticated channel, so fixed role labels are fine.
    private val ID_A = "cc-nearby-A".toByteArray(StandardCharsets.US_ASCII)
    private val ID_B = "cc-nearby-B".toByteArray(StandardCharsets.US_ASCII)

    const val STATUS_REJECT = 0
    const val STATUS_OK = 1
    const val STATUS_PENDING = 2

    /** Identity exchanged on success: enough to create a real internet-messaging contact + TOFU-pin. */
    data class NearbyIdentity(
        val uid: String,
        val publicKeyBase64: String,
        val displayName: String
    )

    /** Maps a phone number to the SPAKE2 password scalar w. Must be identical across platforms. */
    fun deriveW(e164: String): BigInteger {
        val h = MessageDigest.getInstance("SHA-512")
            .digest("$DOMAIN:w:${e164.trim()}".toByteArray(StandardCharsets.UTF_8))
        return BigInteger(1, h).mod(Spake2P256.ORDER)
    }

    /** The searcher: knows the target's number (→ [w]) and wants to add them. */
    class Initiator(private val w: BigInteger, private val me: NearbyIdentity) {
        private val x = Spake2P256.randomScalar()
        private val pA = Spake2P256.shareA(w, x)
        private var ke: ByteArray? = null

        fun message1(): ByteArray = encodeMsg1(pA)

        /** Verifies the responder's confirmation, then returns message 3 (our confirmation + identity). */
        fun onMessage2(msg2: ByteArray): ByteArray {
            val (pB, cB) = decodeMsg2(msg2)
            val k = Spake2P256.keyA(w, x, pB)
            val tt = Spake2P256.transcript(ID_A, ID_B, pA, pB, k, w)
            val keys = Spake2P256.deriveKeys(tt)
            require(constantTimeEquals(cB, Spake2P256.confirm(keys.confirmKeyB, tt))) {
                "Responder confirmation failed (wrong number / MITM)."
            }
            ke = keys.sharedKey
            val cA = Spake2P256.confirm(keys.confirmKeyA, tt)
            return encodeMsg3(cA, encryptIdentity(keys.sharedKey, me))
        }

        /** Reads the final message and returns the peer's identity (or throws on reject). */
        fun onMessage4(msg4: ByteArray): NearbyIdentity {
            val key = ke ?: error("onMessage2 must run first.")
            val (status, enc) = decodeMsg4(msg4)
            require(status == STATUS_OK) { "Pairing was not accepted (status=$status)." }
            return decryptIdentity(key, enc)
        }

        /** Symmetric key for offline Bluetooth P2P chat with this peer, derived from the SPAKE2 secret. */
        fun contactKey(): ByteArray = deriveContactKey(ke ?: error("Handshake not complete."))
    }

    /** The discoverable device: uses its OWN number as the password. */
    class Responder(private val w: BigInteger, private val me: NearbyIdentity) {
        private val y = Spake2P256.randomScalar()
        private var ke: ByteArray? = null
        private var confirmKeyA: ByteArray? = null
        private var tt: ByteArray? = null

        /** Consumes message 1 and returns message 2 (our share + confirmation). */
        fun onMessage1(msg1: ByteArray): ByteArray {
            val pA = decodeMsg1(msg1)
            val pB = Spake2P256.shareB(w, y)
            val k = Spake2P256.keyB(w, y, pA)
            val transcript = Spake2P256.transcript(ID_A, ID_B, pA, pB, k, w)
            val keys = Spake2P256.deriveKeys(transcript)
            ke = keys.sharedKey
            confirmKeyA = keys.confirmKeyA
            tt = transcript
            return encodeMsg2(pB, Spake2P256.confirm(keys.confirmKeyB, transcript))
        }

        /**
         * Verifies the initiator's confirmation and returns their identity. Does NOT build the reply
         * or add a contact — the service layer decides that after user consent (see [responseMessage]).
         */
        fun onMessage3(msg3: ByteArray): NearbyIdentity {
            val key = ke ?: error("onMessage1 must run first.")
            val (cA, enc) = decodeMsg3(msg3)
            require(constantTimeEquals(cA, Spake2P256.confirm(confirmKeyA!!, tt!!))) {
                "Initiator confirmation failed (wrong number / MITM)."
            }
            return decryptIdentity(key, enc)
        }

        /**
         * Message 4. Pass [status] = [STATUS_OK] with the user's consent to reveal our identity and
         * complete the add; [STATUS_PENDING] while awaiting consent; [STATUS_REJECT] to decline.
         */
        fun responseMessage(status: Int): ByteArray {
            val enc = if (status == STATUS_OK) encryptIdentity(ke!!, me) else ByteArray(0)
            return encodeMsg4(status, enc)
        }

        /** Symmetric key for offline Bluetooth P2P chat with this peer, derived from the SPAKE2 secret. */
        fun contactKey(): ByteArray = deriveContactKey(ke ?: error("Handshake not complete."))
    }

    /** A standalone REJECT reply (message 4) for when we can't or won't run the handshake. */
    fun rejectMessage(): ByteArray = encodeMsg4(STATUS_REJECT, ByteArray(0))

    /** Reads just the status byte of a message-4 reply (for the connector's PENDING polling loop). */
    fun responseStatus(msg4: ByteArray): Int = runCatching { decodeMsg4(msg4).first }.getOrDefault(-1)

    private fun deriveContactKey(ke: ByteArray): ByteArray =
        Crypto.hkdfSha256(ke, info = "$DOMAIN:btkey".toByteArray(StandardCharsets.UTF_8), outputLength = 32)

    // ---- Identity encryption (AES-256-GCM under a key derived from the SPAKE2 secret Ke) ----

    private fun identityKey(ke: ByteArray): ByteArray =
        Crypto.hkdfSha256(ke, info = "$DOMAIN:identity".toByteArray(StandardCharsets.UTF_8), outputLength = 32)

    private fun encryptIdentity(ke: ByteArray, identity: NearbyIdentity): ByteArray {
        val plaintext = ByteArrayOutputStream().apply {
            writeLenPrefixed(this, identity.uid.toByteArray(StandardCharsets.UTF_8), 128)
            writeLenPrefixed(this, identity.publicKeyBase64.toByteArray(StandardCharsets.US_ASCII), 512)
            writeLenPrefixed(this, identity.displayName.toByteArray(StandardCharsets.UTF_8), 128)
        }.toByteArray()
        val nonce = Crypto.randomBytes(12)
        val ct = Crypto.aesGcmEncrypt(key = identityKey(ke), nonce = nonce, plaintext = plaintext, associatedData = null)
        return ByteArrayOutputStream().apply {
            write(nonce)
            write((ct.size shr 8) and 0xFF); write(ct.size and 0xFF)
            write(ct)
        }.toByteArray()
    }

    private fun decryptIdentity(ke: ByteArray, enc: ByteArray): NearbyIdentity {
        val reader = Reader(enc)
        val nonce = reader.bytes(12)
        val ctLen = (reader.u8() shl 8) or reader.u8()
        val ct = reader.bytes(ctLen)
        val plaintext = Crypto.aesGcmDecrypt(key = identityKey(ke), nonce = nonce, ciphertextAndTag = ct, associatedData = null)
        val pr = Reader(plaintext)
        val uid = String(pr.lenPrefixed(128), StandardCharsets.UTF_8)
        val pub = String(pr.lenPrefixed(512), StandardCharsets.US_ASCII)
        val name = String(pr.lenPrefixed(128), StandardCharsets.UTF_8)
        return NearbyIdentity(uid = uid, publicKeyBase64 = pub, displayName = name)
    }

    // ---- Message codecs ----

    private fun encodeMsg1(pA: ByteArray): ByteArray =
        ByteArrayOutputStream().apply { write(VERSION); write(pA) }.toByteArray()

    private fun decodeMsg1(bytes: ByteArray): ByteArray {
        val r = Reader(bytes)
        require(r.u8() == VERSION) { "Unsupported version." }
        return r.bytes(65)
    }

    private fun encodeMsg2(pB: ByteArray, cB: ByteArray): ByteArray =
        ByteArrayOutputStream().apply { write(VERSION); write(pB); write(cB) }.toByteArray()

    private fun decodeMsg2(bytes: ByteArray): Pair<ByteArray, ByteArray> {
        val r = Reader(bytes)
        require(r.u8() == VERSION) { "Unsupported version." }
        return r.bytes(65) to r.bytes(32)
    }

    private fun encodeMsg3(cA: ByteArray, encIdentity: ByteArray): ByteArray =
        ByteArrayOutputStream().apply {
            write(VERSION); write(cA)
            write((encIdentity.size shr 8) and 0xFF); write(encIdentity.size and 0xFF)
            write(encIdentity)
        }.toByteArray()

    private fun decodeMsg3(bytes: ByteArray): Pair<ByteArray, ByteArray> {
        val r = Reader(bytes)
        require(r.u8() == VERSION) { "Unsupported version." }
        val cA = r.bytes(32)
        val len = (r.u8() shl 8) or r.u8()
        return cA to r.bytes(len)
    }

    private fun encodeMsg4(status: Int, encIdentity: ByteArray): ByteArray =
        ByteArrayOutputStream().apply {
            write(VERSION); write(status and 0xFF)
            write((encIdentity.size shr 8) and 0xFF); write(encIdentity.size and 0xFF)
            write(encIdentity)
        }.toByteArray()

    private fun decodeMsg4(bytes: ByteArray): Pair<Int, ByteArray> {
        val r = Reader(bytes)
        require(r.u8() == VERSION) { "Unsupported version." }
        val status = r.u8()
        val len = (r.u8() shl 8) or r.u8()
        return status to r.bytes(len)
    }

    private fun writeLenPrefixed(out: ByteArrayOutputStream, value: ByteArray, maxLen: Int) {
        require(value.size <= maxLen) { "value too long" }
        out.write((value.size shr 8) and 0xFF)
        out.write(value.size and 0xFF)
        out.write(value)
    }

    private fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
        if (a.size != b.size) return false
        var diff = 0
        for (i in a.indices) diff = diff or (a[i].toInt() xor b[i].toInt())
        return diff == 0
    }

    private class Reader(private val data: ByteArray) {
        private var pos = 0
        fun u8(): Int = data[pos++].toInt() and 0xFF
        fun bytes(n: Int): ByteArray {
            val slice = data.copyOfRange(pos, pos + n); pos += n; return slice
        }
        fun lenPrefixed(maxLen: Int): ByteArray {
            val len = (u8() shl 8) or u8()
            require(len in 0..maxLen) { "length out of range" }
            return bytes(len)
        }
    }
}

File: iOS/Crisis Connect/Features/Messaging/NearbySpakePairing.swift
//
//  NearbySpakePairing.swift
//  Crisis Connect
//
//  iOS port of the Android `NearbySpakePairing.kt` — the transport-agnostic SPAKE2 session for
//  harvest-proof offline "add by phone number". Kept byte-for-byte compatible with Android: the
//  phone number is only the low-entropy SPAKE2 password (never broadcast), and identities are
//  exchanged AES-256-GCM-encrypted under a key derived from the SPAKE2 secret Ke.
//
//  Flow (initiator = searcher who has the target's number; responder = discoverable device, uses
//  its OWN number). Both derive the same w only if they mean the same number:
//    1. initiator → responder : pA
//    2. responder → initiator : pB, cB
//    3. initiator → responder : cA, enc(identityA)
//    4. responder → initiator : status, enc(identityB)
//

import Foundation
import CryptoKit

enum NearbyPairingError: Error {
    case malformed
    case confirmationFailed
    case rejected(Int)
    case notReady
}

struct NearbyIdentity: Equatable {
    let uid: String
    let publicKeyBase64: String
    let displayName: String
}

enum NearbySpakePairing {
    private static let version: UInt8 = 1
    private static let domain = "crisisconnect:nearby:spake2:v1"
    private static let idA = Data("cc-nearby-A".utf8)
    private static let idB = Data("cc-nearby-B".utf8)

    static let statusReject = 0
    static let statusOk = 1
    static let statusPending = 2

    /// Maps a phone number to the SPAKE2 password scalar w. Must match Android exactly.
    static func deriveW(_ e164: String) -> BigUInt {
        let trimmed = e164.trimmingCharacters(in: .whitespacesAndNewlines)
        let hash = Data(SHA512.hash(data: Data("\(domain):w:\(trimmed)".utf8)))
        return BigUInt(bytes: hash).mod(Spake2P256.order)
    }

    /// A standalone REJECT reply (message 4) for when we can't or won't run the handshake.
    static func rejectMessage() -> Data { encodeMsg4(status: statusReject, encIdentity: Data()) }

    /// The status byte of a message-4 reply (for the connector's PENDING polling loop).
    static func responseStatus(_ msg4: Data) -> Int {
        (try? decodeMsg4(msg4).status) ?? -1
    }

    // MARK: - Initiator (searcher)

    final class Initiator {
        private let w: BigUInt
        private let me: NearbyIdentity
        private let x: BigUInt
        private let pA: Data
        private var ke: Data?

        init(w: BigUInt, me: NearbyIdentity) {
            self.w = w
            self.me = me
            self.x = Spake2P256.randomScalar()
            self.pA = Spake2P256.shareA(w: w, x: x)
        }

        func message1() -> Data { encodeMsg1(pA: pA) }

        func onMessage2(_ msg2: Data) throws -> Data {
            let (pB, cB) = try decodeMsg2(msg2)
            let k = Spake2P256.keyA(w: w, x: x, pB: pB)
            let tt = Spake2P256.transcript(idA: idA, idB: idB, pA: pA, pB: pB, k: k, w: w)
            let keys = Spake2P256.deriveKeys(tt)
            guard constantTimeEquals(cB, Spake2P256.confirm(confirmKey: keys.confirmKeyB, tt: tt)) else {
                throw NearbyPairingError.confirmationFailed
            }
            ke = keys.sharedKey
            let cA = Spake2P256.confirm(confirmKey: keys.confirmKeyA, tt: tt)
            return encodeMsg3(cA: cA, encIdentity: try encryptIdentity(keys.sharedKey, me))
        }

        func onMessage4(_ msg4: Data) throws -> NearbyIdentity {
            guard let key = ke else { throw NearbyPairingError.notReady }
            let decoded = try decodeMsg4(msg4)
            guard decoded.status == statusOk else { throw NearbyPairingError.rejected(decoded.status) }
            return try decryptIdentity(key, decoded.encIdentity)
        }

        func contactKey() throws -> Data {
            guard let key = ke else { throw NearbyPairingError.notReady }
            return deriveContactKey(key)
        }
    }

    // MARK: - Responder (discoverable device, uses its own number)

    final class Responder {
        private let w: BigUInt
        private let me: NearbyIdentity
        private let y: BigUInt
        private var ke: Data?
        private var confirmKeyA: Data?
        private var tt: Data?

        init(w: BigUInt, me: NearbyIdentity) {
            self.w = w
            self.me = me
            self.y = Spake2P256.randomScalar()
        }

        func onMessage1(_ msg1: Data) throws -> Data {
            let pA = try decodeMsg1(msg1)
            let pB = Spake2P256.shareB(w: w, y: y)
            let k = Spake2P256.keyB(w: w, y: y, pA: pA)
            let transcript = Spake2P256.transcript(idA: idA, idB: idB, pA: pA, pB: pB, k: k, w: w)
            let keys = Spake2P256.deriveKeys(transcript)
            ke = keys.sharedKey
            confirmKeyA = keys.confirmKeyA
            tt = transcript
            return encodeMsg2(pB: pB, cB: Spake2P256.confirm(confirmKey: keys.confirmKeyB, tt: transcript))
        }

        func onMessage3(_ msg3: Data) throws -> NearbyIdentity {
            guard let key = ke, let kcA = confirmKeyA, let transcript = tt else {
                throw NearbyPairingError.notReady
            }
            let (cA, enc) = try decodeMsg3(msg3)
            guard constantTimeEquals(cA, Spake2P256.confirm(confirmKey: kcA, tt: transcript)) else {
                throw NearbyPairingError.confirmationFailed
            }
            return try decryptIdentity(key, enc)
        }

        func responseMessage(status: Int) throws -> Data {
            let enc: Data
            if status == statusOk {
                guard let key = ke else { throw NearbyPairingError.notReady }
                enc = try encryptIdentity(key, me)
            } else {
                enc = Data()
            }
            return encodeMsg4(status: status, encIdentity: enc)
        }

        func contactKey() throws -> Data {
            guard let key = ke else { throw NearbyPairingError.notReady }
            return deriveContactKey(key)
        }
    }

    private static func deriveContactKey(_ ke: Data) -> Data {
        hkdf(ke, info: "\(domain):btkey")
    }

    // MARK: - Identity encryption (AES-256-GCM under a key derived from Ke)

    private static func identityKey(_ ke: Data) -> Data { hkdf(ke, info: "\(domain):identity") }

    private static func encryptIdentity(_ ke: Data, _ identity: NearbyIdentity) throws -> Data {
        var plaintext = Data()
        appendLenPrefixed(&plaintext, Data(identity.uid.utf8))
        appendLenPrefixed(&plaintext, Data(identity.publicKeyBase64.utf8))
        appendLenPrefixed(&plaintext, Data(identity.displayName.utf8))

        var nonceBytes = Data(count: 12)
        _ = nonceBytes.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 12, $0.baseAddress!) }
        let box = try AES.GCM.seal(
            plaintext,
            using: SymmetricKey(data: identityKey(ke)),
            nonce: try AES.GCM.Nonce(data: nonceBytes)
        )
        let ct = box.ciphertext + box.tag
        var out = Data()
        out.append(nonceBytes)
        out.append(len2(ct.count))
        out.append(ct)
        return out
    }

    private static func decryptIdentity(_ ke: Data, _ enc: Data) throws -> NearbyIdentity {
        var r = Reader(enc)
        let nonce = try r.bytes(12)
        let ctLen = try r.len2()
        let ctTag = try r.bytes(ctLen)
        guard ctTag.count > 16 else { throw NearbyPairingError.malformed }
        let box = try AES.GCM.SealedBox(
            nonce: try AES.GCM.Nonce(data: nonce),
            ciphertext: ctTag.prefix(ctTag.count - 16),
            tag: ctTag.suffix(16)
        )
        let plaintext = try AES.GCM.open(box, using: SymmetricKey(data: identityKey(ke)))
        var pr = Reader(plaintext)
        let uid = String(decoding: try pr.lenPrefixed(), as: UTF8.self)
        let pub = String(decoding: try pr.lenPrefixed(), as: UTF8.self)
        let name = String(decoding: try pr.lenPrefixed(), as: UTF8.self)
        return NearbyIdentity(uid: uid, publicKeyBase64: pub, displayName: name)
    }

    // MARK: - Message codecs (byte-identical to Android)

    private static func encodeMsg1(pA: Data) -> Data { Data([version]) + pA }
    private static func decodeMsg1(_ bytes: Data) throws -> Data {
        var r = Reader(bytes)
        guard try r.u8() == Int(version) else { throw NearbyPairingError.malformed }
        return try r.bytes(65)
    }

    private static func encodeMsg2(pB: Data, cB: Data) -> Data { Data([version]) + pB + cB }
    private static func decodeMsg2(_ bytes: Data) throws -> (Data, Data) {
        var r = Reader(bytes)
        guard try r.u8() == Int(version) else { throw NearbyPairingError.malformed }
        return (try r.bytes(65), try r.bytes(32))
    }

    private static func encodeMsg3(cA: Data, encIdentity: Data) -> Data {
        Data([version]) + cA + len2(encIdentity.count) + encIdentity
    }
    private static func decodeMsg3(_ bytes: Data) throws -> (Data, Data) {
        var r = Reader(bytes)
        guard try r.u8() == Int(version) else { throw NearbyPairingError.malformed }
        let cA = try r.bytes(32)
        let len = try r.len2()
        return (cA, try r.bytes(len))
    }

    private static func encodeMsg4(status: Int, encIdentity: Data) -> Data {
        Data([version, UInt8(status & 0xFF)]) + len2(encIdentity.count) + encIdentity
    }
    private static func decodeMsg4(_ bytes: Data) throws -> (status: Int, encIdentity: Data) {
        var r = Reader(bytes)
        guard try r.u8() == Int(version) else { throw NearbyPairingError.malformed }
        let status = try r.u8()
        let len = try r.len2()
        return (status, try r.bytes(len))
    }

    // MARK: - Helpers

    private static func hkdf(_ ke: Data, info: String) -> Data {
        HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: ke),
            salt: Data(repeating: 0, count: 32),
            info: Data(info.utf8),
            outputByteCount: 32
        ).withUnsafeBytes { Data($0) }
    }

    private static func len2(_ n: Int) -> Data { Data([UInt8((n >> 8) & 0xFF), UInt8(n & 0xFF)]) }

    private static func appendLenPrefixed(_ out: inout Data, _ value: Data) {
        out.append(len2(value.count))
        out.append(value)
    }

    private static func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
        guard a.count == b.count else { return false }
        var diff: UInt8 = 0
        let ab = [UInt8](a), bb = [UInt8](b)
        for i in 0..<ab.count { diff |= ab[i] ^ bb[i] }
        return diff == 0
    }

    private struct Reader {
        private let data: [UInt8]
        private var pos = 0
        init(_ d: Data) { self.data = [UInt8](d) }
        mutating func u8() throws -> Int {
            guard pos < data.count else { throw NearbyPairingError.malformed }
            defer { pos += 1 }
            return Int(data[pos])
        }
        mutating func bytes(_ n: Int) throws -> Data {
            guard pos + n <= data.count, n >= 0 else { throw NearbyPairingError.malformed }
            defer { pos += n }
            return Data(data[pos..<pos + n])
        }
        mutating func len2() throws -> Int { (try u8() << 8) | (try u8()) }
        mutating func lenPrefixed() throws -> Data { try bytes(try len2()) }
    }
}
